File: src/main/java/de/schramm/royalbash/controller/ExternalModel.java
package de.schramm.royalbash.controller;

import io.swagger.annotations.ApiModelProperty;
import lombok.Builder;
import lombok.Singular;
import lombok.Value;

import java.util.List;
import java.util.stream.Collectors;

@Value
@Builder
class ExternalModel {

    @Value
    @Builder
    static class Game {
        @ApiModelProperty(required = true)
        private Player player1;
        @ApiModelProperty(required = true)
        private Player player2;
        @ApiModelProperty(required = true)
        private String playerOnTurn;
        @ApiModelProperty(required = true)
        private String state;

        static Game toExternalModel(de.schramm.royalbash.controller.service.core.Game game) {
            return Game.builder()
                    .player1(Player.toExternalModel(game.getPlayer1()))
                    .player2(Player.toExternalModel(game.getPlayer2()))
                    .playerOnTurn(game.getPlayerOnTurn().getId())
                    .state(game.getState().name())
                    .build();
        }
    }

    @Value
    @Builder
    static class Player {
        @ApiModelProperty(required = true)
        private String id;
        private String name;
        @ApiModelProperty(required = true)
        private int hitpoints;
        @ApiModelProperty(required = true)
        private int resources;
        @ApiModelProperty(required = true)
        @Singular("deckcard")
        private List<Card> deckcards;
        @ApiModelProperty(required = true)
        @Singular("handcard")
        private List<Card> handcards;
        @ApiModelProperty(required = true)
        @Singular("spot")
        private List<Spot> spots;

        static Player toExternalModel(de.schramm.royalbash.controller.service.core.Player player) {
            return Player.builder()
                    .id(player.getId())
                    .name(player.getName())
                    .hitpoints(player.getHitpoints())
                    .resources(player.getResources())
                    .deckcards(player.getDeckcards()
                            .map(Card::toExternalModel)
                            .collect(Collectors.toList()))
                    .handcards(player.getHandcards()
                            .map(Card::toExternalModel)
                            .collect(Collectors.toList()))
                    .spots(player.getSpots()
                        .map(Spot::toExternalModel)
                        .collect(Collectors.toList()))
                    .build();
        }
    }

    @Value
    @Builder
    static class Card {
        @ApiModelProperty(required = true)
        private String id;
        @ApiModelProperty(required = true)
        private int cost;
        @ApiModelProperty(required = true)
        private boolean placeableOnSpot;

        static Card toExternalModel(de.schramm.royalbash.controller.service.core.Card card) {
            return Card.builder()
                    .id(card.getId())
                    .cost(card.getCost())
                    .placeableOnSpot(card.isPlaceableOnSpot())
                    .build();
        }
    }

    @Value
    @Builder
    static class Spot {
        private Creature creature;

        static Spot toExternalModel(de.schramm.royalbash.controller.service.core.Spot spot) {
            return Spot.builder()
                    .creature(spot
                            .getCreature()
                            .map(Creature::toExternalModel)
                            .orElse(null))
                    .build();
        }
    }

    @Value
    @Builder
    static class Creature {
        @ApiModelProperty(required = true)
        private String id;
        private String name;
        @ApiModelProperty(required = true)
        private int hitpoints;
        @ApiModelProperty(required = true)
        private int atttack;

        static Creature toExternalModel(de.schramm.royalbash.controller.service.core.Creature creature) {
            return Creature.builder()
                    .id(creature.getId())
                    .name(creature.getName())
                    .hitpoints(creature.getHitpoints())
                    .atttack(creature.getAttack())
                    .build();
        }
    }
}

File: src/main/java/de/schramm/royalbash/controller/ExternalModel.kt
package de.schramm.royalbash.controller

import io.swagger.annotations.ApiModelProperty


internal class ExternalModel {

    internal class Game(
            @ApiModelProperty(required = true) val id: String,
            @ApiModelProperty(required = true) val player1: Player,
            @ApiModelProperty(required = true) val player2: Player,
            @ApiModelProperty(required = true) val playerOnTurn: String,
            @ApiModelProperty(required = true) val state: String
    ) {
        companion object {

            fun toExternalModel(game: de.schramm.royalbash.controller.service.core.Game): Game {
                return Game(
                    id = game.id,
                    player1 = Player.toExternalModel(game.player1),
                    player2 = Player.toExternalModel(game.player2),
                    playerOnTurn = game.playerOnTurn.id,
                    state = game.state.name)
            }
        }
    }

    internal class Player(
            @ApiModelProperty(required = true) val id: String,
            @ApiModelProperty(required = true) val name: String = "",
            @ApiModelProperty(required = true) val hitpoints: Int = 0,
            @ApiModelProperty(required = true) val resources: Int = 0,
            @ApiModelProperty(required = true) val deckcards: List<Card> = listOf(),
            @ApiModelProperty(required = true) val handcards: List<Card> = listOf(),
            @ApiModelProperty(required = true) val spots: List<Spot> = listOf()
    ) {

        companion object {

            fun toExternalModel(player: de.schramm.royalbash.controller.service.core.Player): Player {
                return Player(
                        id = player.id,
                        name = player.name,
                        hitpoints = player.hitpoints,
                        resources = player.resources,
                        deckcards = player.deckcards.map { Card.toExternalModel(it) },
                        handcards = player.handcards.map { Card.toExternalModel(it) },
                        spots = player.spots.map {Spot.toExternalModel(it) })
            }
        }
    }

    internal class Card(
            @ApiModelProperty(required = true) val id: String,
            @ApiModelProperty(required = true) val name: String,
            @ApiModelProperty(required = true) val text: String,
            @ApiModelProperty(required = true) val cost: Int = 0,
            @ApiModelProperty(required = true) val placeableOnSpot: Boolean
    ) {

        companion object {

            fun toExternalModel(card: de.schramm.royalbash.controller.service.core.Card): Card {
                return Card(
                        id = card.id,
                        name = card.name,
                        text = card.text,
                        cost = card.cost,
                        placeableOnSpot = card.placeableOnSpot)
            }
        }
    }

    internal class Spot(@ApiModelProperty(required = false) val creature: Creature? = null) {

        companion object {

            fun toExternalModel(spot: de.schramm.royalbash.controller.service.core.Spot): Spot {
                return Spot( creature = spot.getCreature()
                        .map { Creature.toExternalModel(it) }
                        .orElse(null))
            }
        }
    }

    internal class Creature(
            @ApiModelProperty(required = true) val id: String,
            @ApiModelProperty(required = true) val name: String = "",
            @ApiModelProperty(required = true) val hitpoints: Int = 0,
            @ApiModelProperty(required = true) val atttack: Int = 0
    ) {

        companion object {

            fun toExternalModel(creature: de.schramm.royalbash.controller.service.core.Creature): Creature {
                return Creature(
                        id = creature.id,
                        name = creature.name,
                        hitpoints = creature.hitpoints,
                        atttack = creature.attack)
            }
        }
    }
}
